src/inmanta/util/__init__.py

```diff
@@ -119,15 +119,15 @@ def ensure_future_and_handle_exception(
     """Fire off a coroutine from the ioloop thread and log exceptions to the logger with the message"""
     future = ensure_future(action)
 
-    def handler(future):
+    def handler(task: Task[T]) -> None:
         try:
-            exc = future.exception()
+            exc = task.exception()
             if exc is not None:
                 logger.exception(msg, exc_info=exc)
         except CancelledError:
             pass
         finally:
-            notify_done_callback(future)
+            notify_done_callback(task)
```

Comment:
This is tough to review. The typing in `asyncio.tasks` and `asyncio.futures` looks incomplete and even somewhat inconsistent to me, but as far as I can tell this is accurate. Just for the sake of clarity, could you add `future: Task[T]` where we call `ensure_future`?

Reply:
And perhaps on the other lines where we call it as well?

Reply:
good idea, this is something that changed after we wrote the first version of this. Now this returns a task

Reply:
> Now this returns a task

It's not entirely clear to me to be honest:

```python
_FT = TypeVar("_FT", bound=Future[Any])

@overload
def ensure_future(coro_or_future: _FT, *, loop: AbstractEventLoop | None = None) -> _FT: ...  # type: ignore[overload-overlap]
@overload
def ensure_future(coro_or_future: Awaitable[_T], *, loop: AbstractEventLoop | None = None) -> Task[_T]: ...

class Task(Future[_T_co]):
```

Technically I think that if you pass it a non-task `Future`, you get the same type of `Future` back. Any other awaitable (including `Task`) will result in a `Task`. So it may actually be a non-task future, but only if `action` is a non-task future, which can't really be expressed in the typing.

All this said, I think typing it as `Task` is a definite step forward, even if there are some edge cases where it might not (and I'm not even sure about this) be completely accurate.

Reply:
I based myself on the documentation:

https://docs.python.org/3/library/asyncio-future.html#asyncio.ensure_future

Reply:
That's the same, isn't it? The first bullet point might result in a non-task `Future`.

Reply:
indeed. What do we do here? Do we type it as `Task | Future`? We can also try to type the method as `Task[T]` so that we never pass a future into it?

Reply:
I did a quick check and according to mypy we only pass a coroutine. So a solution might be to type it as: `Coroutine[Any, Any, T]` instead of `Awaitable[T]`

I think there are a number of options:

1. We type it correct that the return type if `Task[T] | Future[T]` when the argument is `Awaitable[T]`. I am not a big fan of this because we never use a future (as far as I can tell).
2. We keep the typing like it is now
3. We keep the typing like it is but add an assert `assert not asyncio.isfuture(future)` (with a comment because the naming sucks). This is risky because we create a runtime failure where we shouldn't (we can also wrap it in an if typingchecking)
4. We change the type of the method to `Coroutine[Any, Any, T]` instead of `Awaitable[T]` In that case ensure_future will always return a Task

I seem to lean towards 4 because that means we can also start moving to https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task and use task groups for cancellation.

For 4 I would move it to another PR though.

Reply:
4 sounds ok to me. Though I'd think that we'd want to use `object` instead of `Any`?

Reply:
I used any because I was not sure if object is correct here. The types are not relevant for us. I can try and see what it does, I am just a bit concerned that if we pass a coroutine from an external library. But everything is a subclass of object so it should work.

Reply:
Perhaps `Coroutine[object, Never, T]`? (`Couritine` is `Generic[YieldType, SendType, ReturnType]`) so I suspect the first is covariant while the second is contravariant.

Reply:
Confirmed: https://github.com/python/typeshed/blob/994d22bfbb542956be24580323d5b4aeef822c77/stdlib/typing.pyi#L406

Reply:
Why would you use Never? It might be that the code that wants to background the coroutine does need a send type

Reply:
How so? We used plain `Awaitable` before so we're clearly not sending anything.

Reply:
Coroutine is a subclass of Awaitable. Say we use a coroutine in a service to which we send something and then we background it. In utils we are not sending something, but the type would not be Coroutine[object, Never, T].

Reply:
Except that the inferred type for `async`-based coroutines seems to use `Any`, so we really have no way to know... The pains of gradual typing...

Reply:
I now have a change that uses Coroutine[object, object, T], would that work for you?

Reply:
`asyncio` does send things: https://github.com/python/cpython/blob/2afc7182e66635b3ec7efb59d2a6c18a7ad1f215/Lib/asyncio/tasks.py#L301

But as far as I can tell it only sends `None`, so we could make it `Couroutine[object, None, T]`?

Reply:
ok, I will try that. We can always change it later if it does not work out

Reply:
> would that work for you?

I'm afraid that's always too restrictive. When I proposed it I didn't take the contra-variance into account. It would require all coroutines to be able to handle any kind of object you send them. I have no clue if they do.

Reply:
so that means any again?